blog/content/post/2025-07-xx-edk2-on-odroid-m2.md

```diff
@@ -0,0 +1,325 @@
+  RK8XX_VOLTAGE_INIT (MASTER_NLDO2,  850000),
+  /* The OPi is officially configured for the 837500 voltage, but is still marked as avdd_0v75_s0 in the schematic and Linux device tree. rockchip says this voltage is set to improve HDMI stability. */
+  RK8XX_VOLTAGE_INIT (MASTER_NLDO3,  837500),
+  RK8XX_VOLTAGE_INIT (MASTER_NLDO4,  850000),
+  RK8XX_VOLTAGE_INIT (MASTER_NLDO5,  750000),
+
+  RK8XX_VOLTAGE_INIT (MASTER_PLDO1,  1800000),
+  RK8XX_VOLTAGE_INIT (MASTER_PLDO2,  1800000),
+  RK8XX_VOLTAGE_INIT (MASTER_PLDO3,  1200000),
+  RK8XX_VOLTAGE_INIT (MASTER_PLDO4,  3300000),
+  RK8XX_VOLTAGE_INIT (MASTER_PLDO5,  3300000),
+  RK8XX_VOLTAGE_INIT (MASTER_PLDO6,  1800000),
+  /* No dual PMICs on this platform */
+};
+```
+
+vdd_0v75_s3: nldo-reg1 {
+  regulator-name = "vdd_0v75_s3";
+  regulator-always-on;
+  regulator-boot-on;
+  regulator-min-microvolt = <750000>;
+  regulator-max-microvolt = <750000>;
+
+  regulator-state-mem {
+    regulator-on-in-suspend;
+    regulator-suspend-microvolt = <750000>;
+  };
+};
+```
+
+According to this snippet, the NLDO1 regulator is powering the vdd_0v75_s3
+plane, and the voltage is 0.75V.
+
+Repeat this for every voltage regulator in the devicetree and we get this:
+
+```c
+static struct regulator_init_data rk806_init_data[] = {
+    /* Master PMIC */
+    RK8XX_VOLTAGE_INIT(MASTER_BUCK1, 950000),
+    RK8XX_VOLTAGE_INIT(MASTER_BUCK2, 950000),
+    RK8XX_VOLTAGE_INIT(MASTER_BUCK3, 750000),
+    RK8XX_VOLTAGE_INIT(MASTER_BUCK4, 950000),
+    RK8XX_VOLTAGE_INIT(MASTER_BUCK5, 900000),
+    /* This is not configured in the M2's Linux device tree
+    RK8XX_VOLTAGE_INIT(MASTER_BUCK6, 1100000), */
+    RK8XX_VOLTAGE_INIT(MASTER_BUCK7, 2000000),
+    RK8XX_VOLTAGE_INIT(MASTER_BUCK8, 3300000),
+    RK8XX_VOLTAGE_INIT(MASTER_BUCK10, 1800000),
+
+    RK8XX_VOLTAGE_INIT(MASTER_NLDO1, 750000),
+    RK8XX_VOLTAGE_INIT(MASTER_NLDO2, 900000),
+    RK8XX_VOLTAGE_INIT(MASTER_NLDO3, 837500),
+    RK8XX_VOLTAGE_INIT(MASTER_NLDO4, 850000),
+    /* RK8XX_VOLTAGE_INIT(MASTER_NLDO5, 750000),*/
+
+PcieIoInit (
+  UINT32  Segment
+  )
+{
+  /* Set reset to gpio output mode */
+  if (Segment == PCIE_SEGMENT_PCIE20L2) {
+    // M.2 M Key
+    GpioPinSetDirection (3, GPIO_PIN_PD1, GPIO_PIN_OUTPUT);
+  }
+}
+
+VOID
+EFIAPI
+PciePowerEn (
+  UINT32   Segment,
+  BOOLEAN  Enable
+  )
+{
+  /* nothing to power on */
+}
+
+VOID
+EFIAPI
+PciePeReset (
+  UINT32   Segment,
+  BOOLEAN  Enable
+  )
+{
+  if (Segment == PCIE_SEGMENT_PCIE20L2) {
+    GpioPinWrite (3, GPIO_PIN_PD1, !Enable);
+  }
+}
+```
+
+From this we can gather a couple of things:
+
+- There is a M.2 M-key slot on the board
+- There is no Power Enable signal for PCIe on this board
+- There is a PCIe Reset signal, in this case connected to GPIO_PIN_PD1 on GPIO
+  controller 3
+
+We need to understand how all this relates to the Odroid M2 we're working with.
+Thankfully, Hardkernel provides us with schematics for this board.
+
+[ insert schematic here ]
```

Comment:
missing?

Reply:
https://github.com/3mdeb/news-and-ideas/pull/674/commits/237086f700e6cefdbc499221085d46524150c1e8